### PR TITLE
Validate ROI in imgcodec

### PR DESCRIPTION
**Pull request by staniewzki**
Signed-off-by: Michał Staniewski <m.staniewzki@gmail.com>

## Category:
**New feature** (*non-breaking change which adds functionality*)

## Description:

Currently, ROI passed to decoders is not validated in any place. The validation should happen at the higher level, as the output shape must also be calculated there, so there are no duplicate calculations and duplicate code. The `OutputShape` function is meant to calculate the shape at the higher level, so adding the ROI validation there is sufficient.
While browsing the code related to shapes in the decoders, I found few small adjustments that could be made, so I have added them here.

### Key points relevant for the review:

N/A

### Tests:

- [x] Existing tests apply
- [x] New tests added
  - [ ] Python tests
  - [x] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [x] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [ ] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

File: dali/imgcodec/util/output_shape.h [hunk 1]
```diff
@@ -44,11 +44,19 @@ void OutputShape(OutShape &out_shape,
     if (rotate) {
       in_d = 1 - d;
     }
+
     int extent = info.shape[in_d];
-    if (d < roi.end.size())
+    if (d < roi.end.size()) {
+      DALI_ENFORCE(0 <= roi.end[d] && roi.end[d] <= info.shape[in_d],
+                   "ROI end must fit into the output image shape");
```
Comment (jantonguirao) [inline dali/imgcodec/util/output_shape.h:51]:
```suggestion
                   "ROI end must fit within the image bounds");
```

File: dali/imgcodec/util/output_shape.h [hunk 2]
```diff
@@ -44,11 +44,19 @@ void OutputShape(OutShape &out_shape,
...
       extent = roi.end[d];
-    if (d < roi.begin.size())
+    }
+    if (d < roi.begin.size()) {
+      DALI_ENFORCE(0 <= roi.begin[d] && roi.begin[d] <= info.shape[in_d],
+                   "ROI begin must fit into the output image shape");
```
Comment (jantonguirao) [inline dali/imgcodec/util/output_shape.h:56]:
```suggestion
                   "ROI begin must fit within the image bounds");
```

**Review by jantonguirao — APPROVED** [review]

**Comment by staniewzki** [comment]
!build

**Review by awolant — APPROVED** [review]

**Comment by staniewzki** [comment]
!build

**Comment by staniewzki** [comment]
!build

**Comment by jantonguirao** [comment]
!build

**Pull request merged**